### ISSUE TITLE
Existance assertions on known existing keys do not raise paradoxes

### ISSUE THREAD

**Issue by danog** [vimeo/psalm#6924]
https://psalm.dev/r/d7d9e69db0

**Comment by danog** [comment]
Using a variable as key actually raises a (very unclear) assertion, only for `array_key_exists`

**Comment by orklah** [comment]
Those are 3 different assertions.

- array_key_exists check that the key exists. 
- isset checks that the key exists and the value is not null(so not a duplicate)
- literal `$a['key']` assume that the key exists (otherwise notice) and the value is truthy (meaning '0' won't pass the if, but it passed array_key_exists and isset)

**Comment by danog** [comment]
Indeed, and they all do not raise RedundantCondition issues if the key is a literal, and `array_key_exists` raises a very unclear RedundantCondition issue only if the key is contained in a variable.

**Comment by danog** [comment]
Something like this should raise a redundantcondition for all three expressions: https://psalm.dev/r/7851b7c6af

**Comment by danog** [comment]
More fun: https://psalm.dev/r/2763a4a0ca

It seems like the logic in `getArrayKeyExistsAssertions` is a bit broken, I initially wanted to extend it to raise an issue when a literal key was being used on a non-shaped array, but it seems some additional fixes are required, too

**Comment by danog** [comment]
This seems to be at least partially working in phpstan: https://phpstan.org/r/fa2d9379-b285-4c7c-b914-32138da2cbcc

**Comment by orklah** [comment]
The array_key_exists seem easy at least. In SimpleAssertionReconciler on line 95, is the condition that handles the reconciliation between the existing type and the array-key-exists assertion.

We're currently switching the possibly_undefined flag to false. I guess we could call triggerIssueForImpossible if the flag was already to false

**Comment by danog** [comment]
It seems like the issue isn't exactly an easy problem, if solved properly.
Initially I thought that the logic in `ArrayFetchAnalyzer` had to be extended to account for existing keys; however, after playing around a bit, I realized that it would be way better to implement this the assertion reconciler: however, it looks like the assertion reconciler won't work with types containing literals, only actual `String_` php-parser nodes (#6938).

I think that an `array-key-exists` assertion should also be emitted alongside `empty` and `isset` assertions, resolving eventual offset variables to their respective literal types, and then eventual paradox checking should be made in the SimpleAssertionReconciler.
